Add an archival queue factory

MichaelSnowden opened the pull request:
<!-- Describe what has changed in this PR -->
**What changed?**
Adds an archival queue factory to create the archival queue.

<!-- Tell your future self why have you made these changes -->
**Why?**
I made this change to instantiate the archival queue.

<!-- How have you verified this change? Tested locally? Added a unit test? Checked in staging env? -->
**How did you test it?**
I added a unit test.

<!-- Assuming the worst case, what can be broken when deploying this change to production? -->
**Potential risks**
Not used in production.

<!-- Is this PR a hotfix candidate or require that a notification be sent to the broader community? (Yes/No) -->
**Is hotfix candidate?**
No.

yycptt commented on service/history/archivalQueueFactory.go:
```diff
@@ -0,0 +1,140 @@
+			CheckpointInterval:                  f.Config.ArchivalProcessorUpdateAckInterval,
+			CheckpointIntervalJitterCoefficient: f.Config.ArchivalProcessorUpdateAckIntervalJitterCoefficient,
+			MaxReaderCount:                      f.Config.QueueMaxReaderCount,
+			TaskMaxRetryCount: func() int {
+				// archival queue task will be retried forever
+				return math.MaxInt32
+			},
```
This is actually critical count for logging/metrics, plz don't use max int32.

MichaelSnowden replied:
Fixed

yycptt commented on service/history/archivalQueueFactory.go:
```diff
@@ -0,0 +1,140 @@
+		archivalQueueFactoryParams: params,
+		QueueFactoryBase: QueueFactoryBase{
+			HostScheduler:        hostScheduler,
+			HostPriorityAssigner: queues.NewPriorityAssigner(),
```
plz double check if archival task is marked as low priority.

MichaelSnowden replied:
Done and added a test for it

yycptt commented on service/history/archivalQueueFactory.go:
```diff
@@ -0,0 +1,140 @@
+func NewArchivalQueueFactory(
+	params archivalQueueFactoryParams,
+) QueueFactory {
+	hostScheduler := queues.NewNamespacePriorityScheduler(
```
I am not sure if we need resource isolation for archival queue and have one task channel for each ns. Probably `queues.NewPriorityScheduler` is enough and we just need to provide weight like 10 or 100 for that one channel.

MichaelSnowden replied:
Now using `queues.NewPriorityScheduler` with this:

```go
ArchivalProcessorSchedulerRoundRobinWeights: dc.GetMapProperty(
  dynamicconfig.ArchivalProcessorSchedulerRoundRobinWeights,
  ConvertWeightsToDynamicConfigValue(DefaultActiveTaskPriorityWeight),
)
```

Please let me know if that looks right

MichaelSnowden replied:
^ never mind, I'm just not setting the weights and it looks like they just use the default weights in that case which all other calls to `queues.NewPriorityScheduler` do

yycptt commented on service/history/archivalQueueFactory.go:
```diff
@@ -0,0 +1,140 @@
+		queues.NamespacePrioritySchedulerOptions{
+			WorkerCount:            params.Config.ArchivalProcessorSchedulerWorkerCount,
+			ActiveNamespaceWeights: params.Config.ArchivalProcessorSchedulerRoundRobinWeights,
+			// we don't need standby weights because we only run in the active cluster
```
I am not following, standby namespaces don't perform archival?

MichaelSnowden replied:
My mistake, fixed

yycptt commented on service/history/archivalQueueFactory.go:
```diff
@@ -0,0 +1,130 @@
+	params archivalQueueFactoryParams,
+) QueueFactory {
+	hostScheduler := queues.NewPriorityScheduler(
+		queues.PrioritySchedulerOptions{
```
Let's specify the Weight option here, since all archival tasks are low priority and the default weight for low priority tasks is 1 which will make the weighted round robin kinda in-efficient (since each round will dispatch only one task.). Let's use a higher value 10, 20 etc. exactly value doesn't matter too much.

MichaelSnowden replied:
Set this to 10. The high priority is also 10, although that should never be used.

yycptt replied:
👍 plz move the weight definition to the top of the file.

yycptt commented on service/history/configs/config.go:
```diff
@@ -529,18 +529,18 @@ func NewConfig(dc *dynamicconfig.Collection, numberOfShards int32, isAdvancedVis
+		ArchivalProcessorMaxPollRPS:           dc.GetIntProperty(dynamicconfig.ArchivalProcessorMaxPollRPS, 20),
+		ArchivalProcessorMaxPollHostRPS:       dc.GetIntProperty(dynamicconfig.ArchivalProcessorMaxPollHostRPS, 0),
+		ArchivalProcessorSchedulerWorkerCount: dc.GetIntProperty(dynamicconfig.ArchivalProcessorSchedulerWorkerCount, 512),
+		ArchivalProcessorMaxPollInterval:      dc.GetDurationProperty(dynamicconfig.ArchivalProcessorMaxPollInterval, 1*time.Minute),
```
probably 5 mins. For immediate queues this is only used as a catch all case in case we have bugs in the code that stop task loading.

MichaelSnowden replied:
Updated to five minutes, but this is a scheduled queue now.

yycptt reviewed (approved):
Looks good. 👍  Please land after addressing the comments.

yux0 commented:
nit: archival_queue_task_executor_test.go and archival_queue_task_executor.go

MichaelSnowden commented:
> nit: archival_queue_task_executor_test.go and archival_queue_task_executor.go

Fixed (in the parent branch), and also fixed for the factory in this branch.
